cmdline/compose.1
.\" 
.\" Evolution's compose command line utility manual page.
.\" (C) Ximian, Inc. http://www.ximian.com
.\"
.\" Author:
.\"   Miguel de Icaza (miguel@kernel.org)
.\"
.TH Evolution 1 "Evolution 1.0"
.SH NAME
compose \- invokes the Evolution Mail composer tool.
.SH SYNOPSIS
.PP
.B compose 
[\-\-subject SUBJECT] [\-s SUBJECT] [\-\-cc LIST] [\-c LIST] [\-\-bcc
LIST] [\-b LIST] [\-\-body FNAME] to...
.SH DESCRIPTION
The 
.I compose
program is a command line tool that can be used to invoke the
Evolution Mail composer directly from the command line, and allows for
simple configuration trough a number of flags.
.SH OPTIONS
The following options are supported
.TP
.I \-\-subject SUBJECT, \-s SUBJECT
Both forms are used to specify the subject to be used in the mail
message.
.TP
.I \-\-cc LIST, \-c LIST
Both forms are used to specify a comma separated list of addresses to
CC the message to.
.TP
.I \-\-bcc LIST, \-b LIST
Both forms are used to specify a comma separated list of addresses
that this message will be blind copied to.
.TP
.I \-\-body FNAME
Loads the filename FNAME as the contents for the Evolution Mail
composer. 
.SH OTHERS
If these options are not enough, you might want to look into accessing
the Evolution Mail Composer directly by using CORBA.
.SH INTERFACES
The Evolution Mail Composer supports the
IDL:GNOME/Evolution/Composer:1.0 interface and is implemented as the
component OAFIID:GNOME_Evolution_Mail_Composer.
.SH SEE ALSO
Evolution-Composer.idl
.SH BUGS
If you find bugs in the Evolution groupware suite, please report these
using the \fIbug-buddy\fP program in the GNOME distribution.

NEWS
Evolution 3.62.1 2026-10-09
---------------------------

Bug Fixes:
	eds-I#672 - Calendar: Limit which local files are inlined in components
	M!244 - Composer: De-duplicate attachments in a mailto URI (Jaroslav Jindrak)

Miscellaneous:
	EWebView: Set CORS allow list on the WebKitWebView for known URI-s

Translations:
	Efstathios Iosifidis (el)
	Roy-Magne Mo (nn)

Evolution 3.62.0 2026-09-11
---------------------------

Bug Fixes:
	I#3381 - Composer: De-duplicate inline images before send
	I#3383 - junk-filters: Do not leak the child stdin pipe into concurrent spawns (Benjamin Herrenschmidt)
	I#3386 - Default to not use read-only calendars for reminders and conflict search
	I#3387 - EUIParser: Notify about accelerators moved by an action rename (Martin Monperrus (AI-assisted))
	I#3388 - Mail: Validate clickable preview elements are generated by Evolution
	M!235 - Mail: Remove deprecated SHA1 signature algorithm (Robin Haberkorn)

Miscellaneous:
	docs: Add API index for newly added symbols in 3.62 for e-util

Translations:
	Alan Mortensen (da)
	Aurimas Černius (lt)
	Balázs Úr (hu)
	Baurzhan Muftakhidinov (kk)
	burns (pt_BR)
	Emin Tufan Çetin (tr)
	Guillaume Bernard (fr)
	Jiri Eischmann (cs)
	Kjartan Maraas (nb)

Evolution 3.61.3 2026-08-28
---------------------------

Bug Fixes:
	I#2 - Use gnome-shell-calendar-server for Reminders in GNOME
	I#1593 - Composer: Add 'Remove Formatting' option
	I#2714 - Mail: Do not replace first thread message with the latest
	I#3271 - Composer: Remove sender's CSS color styles in HTML reply
	I#3329 - EMailFormatter: Error text hard to read in dark theme
	I#3336 - EMFolderTree: Cache folder icons for quicker drawing
	I#3367 - EActivityProxy: Hide frame shadow
	I#3369 - MessageList: Add style option for important mail color
	I#3374 - Appointment cannot be saved when an e-mail is attached to it
	I#3375 - Composer: Unsetting text color in HTML does not work
	I#3377 - Mail: Embed file-based images in saved HTML signatures
	M!239 - Fix signature editor window headerbar (Aliaksandr Trush)
	M!240 - CI: Convert the tests into real CTest tests

Miscellaneous:
	flatpak: Return back the libmspack build
	e_util_ignore_accel_for_focused: Ignore accels when a button is focused
	EMenuBar: Set tooltip on a menu button
	SetupBuildFlags.cmake: Add `-Wunused-function`
	test-web-view-jsc: Correct the tests for newer WebKitGTK
	CI: Install an NNTP server for libcamel unit tests
	CI: Create also readonly NNTP group for the unit tests
	e_util_ignore_accel_for_focused: Ignore on GtkTreeView only when its search is on
	test-web-view-jsc: Stop test early when script execution fails
	CI: Enable JMAP for the Cyrus server

Translations:
	Anders Jonsson (sv)
	Asier Saratsua Garmendia (eu)
	Aurimas Černius (lt)
	Ekaterine Papava (ka)
	Jean-Marc TISSIERES (fr)
	Jordi Mas (ca)
	Martin (sl)
	Yuri Chornoivan (uk)

Evolution 3.61.2 2026-07-31
---------------------------

Bug Fixes:
	I#3311 - Fails to import Chinese Outook PST file (charset) ][
	I#3315 - Marcus Bains line is displayed on the wrong day ][
	I#3351 - EMFolderTree: Unexpected scroll after context menu dismiss
	I#3352 - Can't toggle email labels in popover via keyboard
	I#3353 - EShellSearchbar: When changing search type, field contents are erased
	I#3355 - cal_shell_view_transfer_selected: Incorrect data type used when moving events
	I#3358 - EShellSearchbar: Hide Saved Search submenu when none is saved
	M!238 - Fix translation of "Collection Account" menu item (Aliaksandr Trush)

Miscellaneous:
	RSS: Reorder when a new folder flags are set and folder summary created
	Addressbook: Improve accessibility of the card view widgets
	ETreeViewFrame: Set tooltip on toolbar buttons
	flatpak: Workaround autoreconf issue in the flatpak build
	flatpak: Remove libmspack module from the build

Translations:
	Aurimas Aurimas Černius (lt)
	Baurzhan Muftakhidinov (kk)
	burns (pt_BR)
	Ekaterine Papava (ka)
	Emin Tufan Çetin (tr)
	Haolin Xue (zh_CN)
	Martin (sl)
	Ryo Nakano (ja)
	Vasil Pupkin (be)
	Yuri Chornoivan (uk)

Evolution 3.61.1 2026-06-26
---------------------------

Bug Fixes:
	I#1304 - Mail: Don't disable folder tree sensitivity after search setup (Alastair D'Silva)
	I#2820 - Do not offer to create a folder under an RSS account
	I#2833 - Add composer format override per folder, recipient, and account
	I#2839 - Localization is not applied for the date field in the email preview
	I#2894 - Composer open delayed by gpg call with stale gpg lock
	I#3176 - itip-formatter: Guess attendee for REQUEST without attendees
	I#3184 - itip-formatter: Layout has excessive empty space, long content
	I#3192 - Mail: Allow to save emails as .eml via drag & drop
	I#3215 - Composer: Inherit theme color not honoured on send
	I#3232 - Better integration of custom layout dialog
	I#3256 - Composer: Add 'View->Use Theme Colors' for HTML mode (view only)
	I#3265 - prefer-plain: Remote content load reverts shown part
	I#3272 - Blank email body for some S/MIME Encrypted mails
	I#3273 - Incorrect month name declination in some locales
	I#3283 - Paste into 'add/edit note' window not working after open
	I#3287 - EUIAction: Declare alternative label for popup menus
	I#3302 - Truncated file stored in the (mail) cache
	I#3307 - prefer-plain: Empty body when prefer plain with HTML and no plain text
	I#3311 - Fails to import Chinese Outook PST file (charset)
	I#3312 - tnef: Fix OOB read when PR_ATTACH_DATA_OBJ buffer is too short
	I#3313 - tnef: Fix PT_UNICODE property over-read via strlen()
	I#3315 - Marcus Bains line is displayed on the wrong day
	I#3316 - Work Week/Day view overlaps text for multi-day events with times
	I#3317 - Inconsistent timezone for completed tasks
	I#3322 - IMAP: Add "Manage subscriptions" into "Receiving options"
	I#3325 - WebDAVBrowser: Do not save collection color when None is set
	I#3326 - Fix line breaks for <p> and heading elements in HTML-to-markdown
	I#3328 - Mail: Reject message send when account's transport source is disabled

cmdline/compose.c
/*
 * compose.c: A commnand line tool to invoke the Evolution mail composer
 *
 * Author:
 *   Miguel de Icaza (miguel@ximian.com)
 *
 * (C) 2001 Ximian, Inc.
 */
#include <config.h>
#include <gnome.h>
#include <liboaf/liboaf.h>
#include <bonobo.h>
#include "composer/Composer.h"

static char *subject;
static char *cc;
static char *bcc;
static char *body;
static char *to = "";

const struct poptOption compose_popt_options [] = {
	{ "subject", 's', POPT_ARG_STRING,
	  &subject,  0,   N_("Subject for the mail message"), N_("SUBJECT") },
	{ "cc",      'c', POPT_ARG_STRING,
	  &cc,       0,   N_("List of people that will be Carbo Copied"), N_("CC-LIST") },
	{ "bcc",     'b', POPT_ARG_STRING,
	  &bcc,      0,   N_("List of people to Blind Carbon Copy this mail to"), N_("BCC-LIST") },
	{ "body",    0,   POPT_ARG_STRING,
	  &body,     0,   N_("Filename containing the body of the message"), N_("BODY-FILE") },
	{ NULL,      0, 0, NULL, 0 }
};

static void
error (const char *msg)
{
	GtkWidget *dialog;
	
	dialog = gnome_message_box_new (
		msg,
		GNOME_MESSAGE_BOX_ERROR,
		GNOME_STOCK_BUTTON_OK,
		NULL);
	
	gnome_dialog_run_and_close (GNOME_DIALOG (dialog));
	exit (1);
	g_assert_not_reached ();
}

GNOME_Evolution_Composer_RecipientList *
make_list (char *str)
{
	GNOME_Evolution_Composer_RecipientList *list;
	char *p;
	int count = 0;

	if (str == NULL)
		str = "";
	
	list = GNOME_Evolution_Composer_RecipientList__alloc();

	if (*str)
		count = 1;

	for (p = str; *p; p++){
		if (*p == ',')
			count++;
	}
	list->_maximum = count;
	list->_length = count;
	list->_buffer = CORBA_sequence_GNOME_Evolution_Composer_Recipient_allocbuf (count);
	
	for (count = 0; (p = strtok (str, ",")) != NULL; count++){
		GNOME_Evolution_Composer_Recipient *x;

		x = GNOME_Evolution_Composer_Recipient__alloc ();

		list->_buffer [count].name = CORBA_string_dup ("");
		list->_buffer [count].address = CORBA_string_dup (p);
		count++;
		str = NULL;
	}

	return list;
}

gint
do_launch (void)
{
	GNOME_Evolution_Composer_RecipientList *to_list, *cc_list, *bcc_list;
	GNOME_Evolution_Composer composer;
	CORBA_Environment ev;

	CORBA_exception_init (&ev);
	composer = bonobo_get_object (
		"OAFIID:GNOME_Evolution_Mail_Composer",
		"GNOME/Evolution/Composer", &ev);
	CORBA_exception_free (&ev);

	if (composer == CORBA_OBJECT_NIL)
		error (_("It was not possible to start up the Evolution Mail Composer"));

	to_list = make_list (to);
	cc_list = make_list (cc);
	bcc_list = make_list (bcc);

	if (subject == NULL)
		subject = "";
	
	GNOME_Evolution_Composer_setHeaders (composer, to_list, cc_list, bcc_list, subject, &ev);
	GNOME_Evolution_Composer_show (composer, &ev);

	return FALSE;
}

int 
main (int argc, char *argv [])
{
	poptContext ctxt = NULL;
	CORBA_ORB orb;
	
	gnomelib_register_popt_table (oaf_popt_options, _("Oaf options"));
	gnome_init_with_popt_table ("Compose", "1.0", argc, argv,
				    compose_popt_options, 0, &ctxt);

	orb = oaf_init (argc, argv);
	if (bonobo_init (NULL, NULL, NULL) == FALSE)
		error (_("It was not possible to initialize the Bonobo component system"));

	if (ctxt){
		const char **to_args = NULL;
		GString *to_str = g_string_new ("");
		int i;
		
		to_args = poptGetArgs (ctxt);

		if (to_args){
			for (i = 0; to_args [i]; i++) {
				if (i > 1)
					g_string_append_c (to_str, ',');
				
				g_string_append (to_str, to_args [i]);
			}
		}
		to = to_str->str;
	}

	gtk_idle_add (GTK_SIGNAL_FUNC (do_launch), NULL);
	
	bonobo_main ();
	
	return 0;
}
